### PR TITLE
Fix/best model checkpoint fix

### LINKED ISSUE #35609: Trainer sets `state.best_model_checkpoint` even when it doesn't save there; leads to training crash
### System Info

- `transformers` version: 4.49.0.dev0
- Platform: Windows-10-10.0.22631-SP0
- Python version: 3.9.16
- Huggingface_hub version: 0.24.7
- Safetensors version: 0.4.5
- Accelerate version: 0.34.2
- Accelerate config:    not found
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA GeForce RTX 3090

### Who can help?

@muellerz 
@SunMarc 
@seanswyi

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

`pytest tests/test_model_card.py::test_model_card` from `setfit` (link: https://github.com/huggingface/setfit/blob/main/tests/test_model_card.py#L15)

Apologies for not having a convenient ready-to-go `transformers`-only script. I'm afraid I don't have time for that right now.
In essence, the flow is as follows:
1. I start the trainer, with lots of evaluations (e.g. `eval_steps=1`, `eval_strategy="steps"`)
2. When evaluating, the new `_determine_best_metric` is called: https://github.com/huggingface/transformers/blob/f63829c87bd89a4a0cea45d81c1cd870996b30c4/src/transformers/trainer.py#L3070-L3075
3. With `args.metric_for_best_model` set, we only set the `best_metric` in the first evaluation: https://github.com/huggingface/transformers/blob/f63829c87bd89a4a0cea45d81c1cd870996b30c4/src/transformers/trainer.py#L3182
4. On the 2nd eval, we start comparing against the first. If the model is better, we now also set `best_model_checkpoint`: https://github.com/huggingface/transformers/blob/f63829c87bd89a4a0cea45d81c1cd870996b30c4/src/transformers/trainer.py#L3184-L3192 *but* we're not sure if we're even going to be saving at this step! If `args.save_strategy != SaveStrategy.BEST:`, then it's very possible that we're not saving.
5. The eventual crash occurs when "deleting old checkpoints", because there is no file at `best_model_checkpoint`: https://github.com/huggingface/transformers/blob/f63829c87bd89a4a0cea45d81c1cd870996b30c4/src/transformers/trainer.py#L2680-L2685

### Expected behavior

We should not be setting `best_model_checkpoint` unless we're confident that 1) `state.should_save` is True or 2) `args.save_strategy == "best"`. Then we'll avoid this crash.

- Tom Aarsen

### PR DESCRIPTION
**Pull request by seanswyi**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/35609 (TL;DR `best_model_checkpoint` is being set when the checkpoint may not even exist)

- Add a new attribute to TrainerState called `best_global_step` that keeps track of the step where we performed evaluation and achieved a new best metric.
- Moved setting `best_model_checkpoint` from `_determine_best_metric` to `_save_checkpoint`.
- Inside `_save_checkpoint`, check if the `best_model_checkpoint` actually exists, and set it if so.

This was a bit trickier than I thought it would be, since the Trainer's saving and evaluation logic are not tightly coupled as would be in the `save_strategy == "best"` scenario.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/35609
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@muellerzr 
@SunMarc 
@tomaarsen 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by SunMarc** [comment]
tests/trainer/test_trainer.py::TrainerIntegrationTest::test_best_model_checkpoint_behavior test you created is failing, can you check ?

**Comment by seanswyi** [comment]
> tests/trainer/test_trainer.py::TrainerIntegrationTest::test_best_model_checkpoint_behavior test you created is failing, can you check ?

I think the error should be fixed now. Problem was I had hardcoded values and wasn't using the patch correctly. Thanks for pointing it out.

**Comment by seanswyi** [comment]
> Make sure to run make style for the CI. There is also a failing test but that's unrelated to this PR

Done. I'm not sure why but there was a bit of discrepancy between my local setting and the CI pipeline so I had to make the changes through a Docker image since `make style` wasn't working.

**Comment by SunMarc** [comment]
thanks ! i will wait for @tomaarsen to test the PR before reviewing it

**Comment by tomaarsen** [comment]
I haven't reviewed the code in this PR, but I can confirm that my test passes with this branch again!
cc @SunMarc 

- Tom Aarsen

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3202,6 +3200,13 @@ def _save_checkpoint(self, model, trial):
         output_dir = os.path.join(run_dir, checkpoint_folder)
         self.save_model(output_dir, _internal_call=True)
 
+        if self.args.save_strategy in [SaveStrategy.STEPS, SaveStrategy.EPOCH] and self.state.best_global_step:
```
Comment (SunMarc) [inline src/transformers/trainer.py:3203]: why don't we consider `SaveStrategy.BEST` also ?

File: src/transformers/testing_utils.py [hunk 2]
```diff
@@ -1413,6 +1415,28 @@ def get_tests_dir(append_path=None):
         return tests_dir
 
 
+def get_steps_per_epoch(trainer: Trainer) -> int:
+    train_dataloader = trainer.get_train_dataloader()
+    batches_per_epoch = len(train_dataloader)
+    steps_per_epoch = math.ceil(batches_per_epoch / trainer.args.gradient_accumulation_steps)
+
+    return steps_per_epoch
```
Comment (SunMarc) [inline src/transformers/testing_utils.py:1423]: use the following to get the wanted value: `set_initial_training_values` in Trainer !

**Review by SunMarc — APPROVED** [review]
Thanks for the PR and sorry for the wait ! Left a few comments !

File: src/transformers/testing_utils.py (hunk 2 above)
Comment (seanswyi) [reply src/transformers/testing_utils.py]: > You can use the following function to get the wanted value: `set_initial_training_values` in Trainer !

Thanks for the update, I don't believe that was available at the time of this PR. Updated the code to use that method instead!

File: src/transformers/trainer.py (hunk 1 above)
Comment (seanswyi) [reply src/transformers/trainer.py:3205]: > why don't we consider `SaveStrategy.BEST` also ?

I purposefully left `SaveStrategy.BEST` out because it seemed unnecessary.

If `save_strategy == SaveStrategy.BEST` then we're guaranteed that evaluation and saving are tightly coupled, implying that `best_model_checkpoint` will always have an existing checkpoint directory since saving always happens after evaluation. However, if `save_strategy in [SaveStrategy.STEPS, SaveStrategy.EPOCH]` saving and evaluation happen separately, meaning that we need to perform manual checks to ensure that FileNotFoundErrors don't occur.

My initial PR that implemented `SaveStrategy.BEST` overlooked this scenario because I didn't consider that saving and evaluation may happen separately, which was why the errors were occurring in the first place.
